keyboards/handwired/ergodox_mass/matrix.c

```diff
@@ -0,0 +1,260 @@
+  L2 = 5
+} GPIOChip_t;
+
+// Forward declarations
+void spi_init(void);
+uint8_t spi_send(uint8_t byte);
+uint8_t gpx_control_byte(GPIOChip_t idx, uint8_t read);
+uint16_t gpx_read_reg(GPIOChip_t idx, uint8_t addr);
+void gpx_write_reg(GPIOChip_t idx, uint8_t addr, uint16_t val);
+
+// GPIO Expander Register Addresses
+#define GPX_IODIR   0x00
+#define GPX_IPOL    0x02
+#define GPX_GPINTEN 0x04
+#define GPX_DEFVAL  0x06
+#define GPX_INTCON  0x08
+#define GPX_IOCON   0x0A
+#define GPX_GPPU    0x0C
+#define GPX_INTF    0x0E
+#define GPX_INTCAP  0x10
+#define GPX_GPIO    0x12
+#define GPX_OLAT    0x14
+
+// Store raw and debounced switch state
+static matrix_row_t raw_matrix[MATRIX_ROWS];
+static matrix_row_t deb_matrix[MATRIX_ROWS];
+
+/**
+ * Define weak empty shells for these functions in case they don't get defined elsewhere
+ */
+
+__attribute__ ((weak))
+void matrix_init_quantum(void) {
+  matrix_init_kb();
+}
+
+__attribute__ ((weak))
+void matrix_scan_quantum(void) {
+  matrix_scan_kb();
+}
+
+__attribute__ ((weak))
+void matrix_init_kb(void) {
+  matrix_init_user();
+}
+
+__attribute__ ((weak))
+void matrix_scan_kb(void) {
+  matrix_scan_user();
+}
+
+__attribute__ ((weak))
+void matrix_init_user(void) {
+}
+
+__attribute__ ((weak))
+void matrix_scan_user(void) {
+}
+
+/**
+ * Define matrix functions
+ */
+
+inline
+uint8_t matrix_rows(void) {
+  return MATRIX_ROWS;
+}
+
+inline
+uint8_t matrix_cols(void) {
+  return MATRIX_COLS;
+}
+
+inline
+void matrix_setup(void) {
+}
+
+inline
+bool matrix_is_modified(void) {
+  return true;
+}
+
+inline
+bool matrix_is_on(uint8_t row, uint8_t col) {
+  return (deb_matrix[row] & ((matrix_row_t)1<col));
+}
+
+inline
+matrix_row_t matrix_get_row(uint8_t row) {
+  return deb_matrix[row];
+}
+
+inline
+void matrix_print(void) {
+}
+
+inline
+void matrix_power_up(void) {
+}
+
+inline
+void matrix_power_down(void) {
+}
+
+/**
+ * Initialize the matrix reading circuitry. Called once at the start
+ */
+void matrix_init(void) {
+  // Initialize all switches to off
+  for (uint8_t i = 0; i < MATRIX_ROWS; ++i) {
+    raw_matrix[i] = 0;
+    deb_matrix[i] = 0;
+  }
+
+  debounce_init(MATRIX_ROWS);
+
+  // Initialize SPI bus
+  spi_init();
+
+  // Initialize all GPIO expanders
+  for (int i = R0; i <= L2; ++i) {
+    gpx_write_reg(i, GPX_IODIR, 0xFFFF); // Inputs
+    gpx_write_reg(i, GPX_IPOL,  0xFFFF); // Flip polarity
+    gpx_write_reg(i, GPX_GPPU,  0xFFFF); // Enable pullups
+    gpx_write_reg(i, GPX_IOCON, 0x0808); // Defaults, but enable hardware addressing
+  }
+
+  matrix_init_quantum();
+}
+
+/**
+ * Read the matrix. Called continuously
+ */
+uint8_t matrix_scan(void) {
+  for (int i = R0; i <= L2; ++i) {
+    // A bank upper 8 bits, B bank lower 8 bits
+    uint16_t gpio = gpx_read_reg(i, GPX_GPIO);
+
+    // Even matrix rows are bank B, odd are bank A
+    raw_matrix[i*2] = gpio & 0xFF;
+    raw_matrix[i*2+1] = (gpio >> 8) & 0xFF;
+  }
+
+  debounce(raw_matrix, deb_matrix, MATRIX_ROWS, true);
```

Comment:
Ideally this should be rewritten to actually detect a change, rather than hard coding it to always assume there is a change.   

The reason for that, is that there is some code that actually checks this, and will prevent stuff from timing out correctly.